server/src/main/java/org/opensearch/gateway/GatewayMetaState.java

```diff
@@ -599,4 +601,105 @@ public void close() throws IOException {
             IOUtils.close(persistenceWriter.getAndSet(null));
         }
     }
+
+    /**
+     * Encapsulates the writing of metadata to a remote store using {@link RemoteClusterStateService}.
+     */
+    public static class RemotePersistedState implements PersistedState {
+
+        private ClusterState lastAcceptedState;
+        private ClusterMetadataMarker lastAcceptedMarker;
+        private final RemoteClusterStateService remoteClusterStateService;
+
+        public RemotePersistedState(final RemoteClusterStateService remoteClusterStateService) {
+            this.remoteClusterStateService = remoteClusterStateService;
+        }
+
+        @Override
+        public long getCurrentTerm() {
+            return lastAcceptedState != null ? lastAcceptedState.term() : 0L;
+        }
+
+        @Override
+        public ClusterState getLastAcceptedState() {
+            if (lastAcceptedMarker != null) {
+                assert lastAcceptedState != null : "Last accepted state is not set";
+                assert lastAcceptedState.metadata().indices().size() == lastAcceptedMarker.getIndices().size()
+                    : "Number of indices in last accepted state and marker are different";
+                lastAcceptedMarker.getIndices().stream().forEach(md -> {
+                    assert lastAcceptedState.metadata().indices().containsKey(md.getIndexName())
+                        : "Last accepted state and marker are not in sync";
+                    assert lastAcceptedState.metadata().indices().get(md.getIndexName()).getIndexUUID().equals(md.getIndexUUID())
+                        : "Last accepted state and marker are not in sync";
+                });
```

Comment:
you can a separate method for all these otherwise it will run the loop everytime in production